### PR TITLE
Visual Studio 2019 generator

### LINKED ISSUE #248: bob project generator for Visual Studio
Hi,

this "issue" is an improvement ticket. I am searching for good ideas how to adapt Visual Studio to bob project to generate a **generic** Visual Studio project like the QtCreator plugin. The best would be, if Visual Studio also uses bob to build the software, debugging should be possible and build errors should be parsed the the source code.

I have no idea how to realize this - especially without using CMake generator.

You are welcome for brainstorming :)

### PR DESCRIPTION
**Pull request by jkloetzke**
The first implementation that works on MSYS2. Full support for VS would require that Bob runs natively on Windows which is not possible (yet).

Fixes #248.

### CODE REVIEW

File: pym/bob/generators/VisualStudio.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+            for i in self.sources: ElementTree.SubElement(g, "ClCompile", {"Include" : str(i)})
+            items.append(ElementTree.tostring(g, encoding="unicode"))
+        # TODO: self.resources
+        if self.dependencies:
+            g = ElementTree.Element("ItemGroup")
+            incPaths = []
+            for i in sorted(self.dependencies):
+                dep = ElementTree.SubElement(g, "ProjectReference",
+                        {"Include" : "..\\{NAME}\\{NAME}.vcxproj".format(NAME=i)})
+                ElementTree.SubElement(dep, "Project").text = "{" + str(projects[i].uuid) + "}"
```
Comment (mahaase) [inline pym/bob/generators/VisualStudio.py]: If there is a module, which is splitted into multiple packages, the projects[i] will crash.
As workaround, I surrounded this lines by try-except and everything is fine.

File: pym/bob/generators/VisualStudio.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+            ElementTree.SubElement(n, "NMakeBuildCommandLine").text = cmd + " -b " + self.packagePath
+            ElementTree.SubElement(n, "NMakeReBuildCommandLine").text = cmd + " -b -f --clean " + self.packagePath
+            # TODO: NMakeCleanCommandLine == "rm -rf dev/{build,dist}"?
+            items.append(ElementTree.tostring(n, encoding="unicode"))
+
+            n = ElementTree.Element("PropertyGroup", {"Condition" : "'$(Configuration)|$(Platform)'=='Checkout+Build|Win32'"})
+            ElementTree.SubElement(n, "NMakeOutput")
+            ElementTree.SubElement(n, "NMakeIncludeSearchPath").text = ";".join(incPaths)
+            ElementTree.SubElement(n, "NMakeBuildCommandLine").text = cmd + " " + self.packagePath
+            ElementTree.SubElement(n, "NMakeReBuildCommandLine").text = cmd + " -f --clean " + self.packagePath
```
Comment (mahaase) [inline pym/bob/generators/VisualStudio.py]: Here I added -v, I guess in QtCreator plugin -v is also default?
But this is something I can do: bring all build-modes from Qt-Creator plugin to here.

File: pym/bob/generators/VisualStudio.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+            g = ElementTree.Element("ItemGroup")
+            for i in self.sources: ElementTree.SubElement(g, "ClCompile", {"Include" : str(i)})
+            items.append(ElementTree.tostring(g, encoding="unicode"))
+        # TODO: self.resources
+        if self.dependencies:
+            g = ElementTree.Element("ItemGroup")
+            incPaths = []
+            for i in sorted(self.dependencies):
+                dep = ElementTree.SubElement(g, "ProjectReference",
+                        {"Include" : "..\\{NAME}\\{NAME}.vcxproj".format(NAME=i)})
```
Comment (mahaase) [inline pym/bob/generators/VisualStudio.py]: Windows has trouble with "::" in filesystem, I replaces here by "__".

File: pym/bob/generators/VisualStudio.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        buildMe = os.path.join(self.destination, "buildme.sh")
+        buildMeCmd = baseBuildMe + " " + buildMe
+        if not self.args.update:
+            self.updateFile(buildMe, self.__generateBuildme(extra))
+
+        solutionProjectList = []
+        solutionProjectConfigs = []
+        for name,project in projects.items():
+            p = os.path.join(self.destination, name)
+            os.makedirs(p, exist_ok=True)
```
Comment (mahaase) [inline pym/bob/generators/VisualStudio.py]: Windows has trouble with "::" in filesystem, I replaces here by "__".

File: pym/bob/generators/VisualStudio.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+            self.updateFile(os.path.join(p, name+".vcxproj"), project.generateProject(projects, buildMeCmd),
+                    encoding="utf-8", newline='\r\n')
+            self.updateFile(os.path.join(p, name+".vcxproj.filters"), project.generateFilters(),
+                    encoding="utf-8", newline='\r\n')
+
+            solutionProjectList.append(SOLUTION_PROJECT_TEMPLATE.format(NAME=name, GUID=str(project.uuid).upper()))
+            solutionProjectConfigs.append("\t\t{{{GUID}}}.Build|x86.ActiveCfg = Build|Win32".format(GUID=str(project.uuid).upper()))
```
Comment (mahaase) [inline pym/bob/generators/VisualStudio.py]: Just for the root-project we have to add "\t\t{{{GUID}}}.Build|x86.Build.0", so the "Projektmappe" get a default project, which will be built by clicking <build "Projektmappe">.

**Comment by jkloetzke** [comment]
Most stuff should be addressed. There are still many TODOs and FIXMEs, though...

File: pym/bob/generators/VisualStudio.py (hunk 1 above)
Comment (mahaase) [reply pym/bob/generators/VisualStudio.py]: fixed!

File: pym/bob/generators/VisualStudio.py (hunk 2 above)
Comment (mahaase) [reply pym/bob/generators/VisualStudio.py]: fixed.

File: pym/bob/generators/VisualStudio.py (hunk 3 above)
Comment (mahaase) [reply pym/bob/generators/VisualStudio.py]: fixed.

File: pym/bob/generators/VisualStudio.py (hunk 4 above)
Comment (mahaase) [reply pym/bob/generators/VisualStudio.py]: fixed

File: pym/bob/generators/VisualStudio.py (hunk 5 above)
Comment (mahaase) [reply pym/bob/generators/VisualStudio.py]: also fixed.

**Comment by mahaase** [comment]
I tested the new fixes - well done!! No problems with "::" anymore. Default project can be built by "Projektmappe erstellen". No crashes while creating the vs project. 👍 

Also the exclude feature works well - important in VS for projects with boost ;-)

The only thing remaining is the project tree, where the source code is just splitted into source and header files.

**Comment by jkloetzke** [comment]
> I tested the new fixes - well done!! No problems with "::" anymore. Default project can be built by "Projektmappe erstellen". No crashes while creating the vs project. 👍
> Also the exclude feature works well - important in VS for projects with boost ;-)

Nice to hear.

> The only thing remaining is the project tree, where the source code is just splitted into source and header files.

That should be fixed too. Hopefully I didn't break anything. 😉 

> btw: vs2019, it is also working with vs2017 so far.

Could be but I guess there may be subtle differences in the generated project. At least CMake has distinct generators for them. As I don't have vs2017 to test I would rather not advertise it officially. At least the toolset has a different version...

**Pull request merged**